d4s2_api/models.py

```diff
@@ -297,3 +302,89 @@ class UserEmailTemplateSet(models.Model):
 
     def __str__(self):
         return 'User Email Template Set user <{}>, set: <{}>'.format(self.user.username, self.email_template_set.name)
+
+
+class S3EndpointManager(models.Manager):
+    def get_by_natural_key(self, url):
+        return self.get(url=url)
+
+
+class S3Endpoint(models.Model):
+    """
+    Defines S3 service provider
+    """
+    objects = S3EndpointManager()
+    url = models.CharField(max_length=255, help_text='URL of S3 service', unique=True)
+
+    def __str__(self):
+        return 'S3 Endpoint url: {}'.format(self.url)
+
+
+class S3UserTypes(object):
+    NORMAL = 0
+    AGENT = 1
+    CHOICES = (
+        (NORMAL, 'Normal'),
+        (AGENT, 'Agent'),
+    )
+
+
+class S3User(models.Model):
+    endpoint = models.ForeignKey(S3Endpoint)
+    s3_id = models.CharField(max_length=255, help_text='S3 user ID (aws_access_key_id)')
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    type = models.IntegerField(choices=S3UserTypes.CHOICES, default=S3UserTypes.NORMAL)
+
+    def get_type_label(self):
+        type_choice = S3UserTypes.CHOICES[self.type]
+        if type_choice:
+            return type_choice[1]
+        return ''
+
+    def __str__(self):
+        return 'S3User s3_id: {} user: {} type: {}'.format(self.s3_id, self.user, self.get_type_label())
+
+    class Meta:
+        unique_together = ('endpoint', 'user')
```

Comment:
Should we also add a unique constraint for `s3_id` + `endpoint`?

Reply:
Since we only have two sets of credentials (currently) I was holding off on that so I could still test against the dev s3 instance.

Reply:
I see - and you need 3 `S3User`s.